### DISCUSSION TITLE
'X' is not defined inside Electron

### DISCUSSION THREAD

**Discussion by MMcKester** [webpack/webpack#16255] [Q&A]
Hi, we are building an Electron app and bumping its dependencies from:

**Webpack:** 4.6 ➡️ 5.7
**Electron:** 14 ➡️ 20
**Angular:** 11 ➡️ 14 (hence the webpack bump)

We adjusted our webpack config and run into issues like these:

```
Uncaught ReferenceError: fs is not defined
...
Uncaught ReferenceError: child_process is not defined
...
Uncaught ReferenceError: worker_threads is not defined
...
```

Our app is a pure offline app and therefore is very loose with security settings (nodeIntegration true, webSecurity off, etc.). Basically, we use the `fs` module within the renderer.

Our `webpack.js` from Webpack 4 has mainly these settings:

```javascript
  target: 'electron-renderer',
  node: {
    fs: 'empty',
  },
  externals: {
    'drivelist': 'require("drivelist")',
    'fsevents': "require('fsevents')",
    'electron-clipboard-ex': "require('electron-clipboard-ex')",
  },
```

Question 1: Electron has a module `fs` that can be included in the renderer. Why did I have to set `fs` to `"empty"` to fix the `'fs' is not defined` error that we are having again with Webpack 5?

Question 2: How can we fix the issue now with Webpack 5? I tried many configurations like adding `fs` to `externals` or using:

```javascript
resolve: {
  fallback: {
    fs: false,
    worker_threads: false
  }
}
```

Thanks a lot for your help!

**Comment by alexander-akait** [comment]
Maybe you can provide small repo (just provide examples of problems), weird, it should work fine

**Reply by MMcKester** [reply]
Hey, I'm happy to put together a repo. I have one more observation to share before that could help identify the issue. My repo is based on this [angular-electron](https://github.com/maximegris/angular-electron) template.

They basically import externals through `window.require` [link to entire code](https://github.com/maximegris/angular-electron/blob/a3eeb6a953d3f178d68378c18478ae2e0fc02132/src/app/core/services/electron/electron.service.ts#L20-L26):

```typescript
    if (this.isElectron) {
      this.fs = window.require('fs');
      ....
```

So in order to call `fs` I need to go through `this.electronService.fs`.

If I instead use the import directly `import fs from 'fs';` in the top of my component, then I receive the error:

    `Module not found: Error: Can't resolve 'fs'`.

My app is an Electron app only, so I don't need dynamic imports. The app will never run in the browser. The `import from` worked with Webpack 4. I am not saying it's the culprit but it is certainly a candidate. I hope my description makes sense. Any help is highly appreciated!

**Reply by alexander-akait** [reply]
`window.require('fs')` and `import fs from 'fs';` are different things... Do you have babel? If yes, do you transpile ESM to commonjs or keep ESM modules?
